### PR TITLE
whitelist cluster-logging-operator metrics for telemetry

### PR DESCRIPTION
**Pull request by pmoogi-redhat**
This PR contains whitelisting of the below cluster-logging-operator metrics for Telemetry.
   "log_logging_info", "log_collector_error_count_total", "log_forwarder_pipeline_info", "log_forwarder_input_info"
    "log_forwarder_output_info","log_collected_bytes_total", "log_logged_bytes_total

- Changes done in manifests/0000_50_cluster-monitoring-operator_04-config.yaml to include the above metrics.
- Reference https://issues.redhat.com/browse/LOG-1926 [JIRA ticket for implementing CLO Metrics]
- https://github.com/openshift/cluster-logging-operator/pull/1264 [CLO Metrics PR]

### CODE REVIEW

**Review by moadz — COMMENTED** [review]
Hey before approving, I would like to understand the implications this has on telemeter/RHOBS load. 

* Where are these metrics produced? (just the CLO or in log forwarding daemons) 
* What cardinality are we expecting with these new metrics? (how many new time series, how the number of # time series scale with number of clusters using this version of OS/CLO)
* Do you have these metrics produced somewhere in a test-cluster so I can examine them?
* What is the intended use of these metrics? (Analytics or support/alerting)

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (pmoogi-redhat) [reply manifests/0000_50_cluster-monitoring-operator_04-config.yaml:655]: @moadz 
Kindly check this document for more details https://docs.google.com/document/d/1s7odpnziZSaAsnFFHgsZs12dFzgXcy9eNJOSO4l1qCI/edit#heading=h.9fn9nvtokvat
1. Except log_logged_bytes_total and log_collected_bytes_total metrics rest are published by cluster-logging-operator
2. log_logged_bytes_total and log_collected_bytes_total  are produced by collector daemon set
3. These two metrics got labels being namespace and podname
4. all info type got fixed value time-series with fixed set of labels in the range of two to six [please check above doc for more details]
5. log_collector_errors_total metric is gauge type - for publishing how many times collector incurred error during reconciling
6. Intended use of these metrics - usage analytics, insights for troubleshooting, appropriately setting default resources limits in the cluster

will share a test-cluster for your further examination. thank you.

**Review by moadz — APPROVED** [review]
LGTM. Thank you for providing me with a test cluster for some due diligence. 

The major cost metrics appear to be: 
`log_collected_bytes_total` 
`log_logged_bytes_total` 

As the cardinality will grow proportional to the number of podname/log path increases. Is it possible that this path will change throughout the lifespan of a pod as logs are rotated? Or is the churn limited to pod restarts?  

I'm satisfied with the justifications presented [here](https://docs.google.com/document/d/1s7odpnziZSaAsnFFHgsZs12dFzgXcy9eNJOSO4l1qCI/edit#) as this metrics list is WIP.

**Comment by alanconway** [comment]
@jan--f @moadz how do we get this merged? Do we need to open a BZ, and if so against what project/component?

**Comment by pmoogi-redhat** [comment]
@jan--f kindly review this PR. Added recording rules for two of the metrics which are collected per container basis.

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (moadz) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:655]: Does this mean we're still forwarding the raw metric (instead of the recorded `node:log_collected_bytes_total:sum`)?

Comment (pmoogi-redhat) [reply manifests/0000_50_cluster-monitoring-operator_04-config.yaml:655]: @moadz cluster:log_logged_bytes_total:sum is a name of recording rule that we have set it in the assets/cluster-monitoring-operator/prometheus-rule.yaml to aggregate raw metric log_logged_bytes_total and log_collected_bytes_total.  I see that i mistakenly given record name as node:.. will correct it. thanks for your review.

**Comment by jan--f** [comment]
@pmoogi-redhat You need to run `make generate` again an add the resulting changes in your commit.

**Comment by pmoogi-redhat** [comment]
> make generate

@jan--f on doing that it took out my changes done on the introduced rules
--- a/assets/cluster-monitoring-operator/prometheus-rule.yaml
+++ b/assets/cluster-monitoring-operator/prometheus-rule.yaml
@@ -30,12 +30,6 @@ spec:
         severity: warning
   - name: openshift-kubernetes.rules
     rules:
-    - expr: sum(log_logged_bytes_total{container="",pod!=""}[5m])
-        BY (cluster)
-      record: cluster:log_logged_bytes_total:sum
-    - expr: sum(log_collected_bytes_total{container="",pod!=""}[5m])
-        BY (cluster)
-      record: cluster:log_collected_bytes_total:sum
     - expr: sum(rate(container_cpu_usage_seconds_total{container="",pod!=""}[5m]))
         BY (pod, namespace)
       record: pod:container_cpu_usage:sum

Where i should be adding prometheus recording rules for my metrics aggregation ? Please let me know as it seems another resource is generating prometheus-rule.yaml when i do make generate

**Comment by pmoogi-redhat** [comment]
@jan--f kindly review this PR and seeking your approval for its merge.

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (jan--f) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:652]: The owner name here differs from what is in `Documentation/data-collection.md`. I think this cause the `generate` test to fail.

File: jsonnet/rules.libsonnet
Comment (jan--f) [inline jsonnet/rules.libsonnet:593]: The rule expression is not valid PromQL.
```suggestion
          expr: 'sum(log_logged_bytes_total{container="",pod!=""}) BY (cluster)',
          record: 'cluster:log_logged_bytes_total:sum',
        },
        {
          expr: 'sum(log_collected_bytes_total{container="",pod!=""}) BY (cluster)',
          record: 'cluster:log_collected_bytes_total:sum',
```
would be valid PromQL. Please double check this rule actually does what you want.

**Review by jan--f — CHANGES_REQUESTED** [review]
Still some issues to address here.

File: jsonnet/rules.libsonnet
Comment (pmoogi-redhat) [reply jsonnet/rules.libsonnet]: resolved

**Comment by jan--f** [comment]
/lgtm
@bburt-rh @sferich888 Please take a look for `docs-approved` and `qe-approved` labels.

**Comment by bburt-rh** [comment]
@jan--f Docs just needs to ack the epic. There's no need for approval of individual PRs. That being said, I'm not sure of the monitoring docs impact, if any, of these changes.

**Comment by jan--f** [comment]
> @jan--f Docs just needs to ack the epic. There's no need for approval of individual PRs. That being said, I'm not sure of the monitoring docs impact, if any, of these changes.

There are probably none, at least from the monitoring side. I just didn't want to preempt you. I'll add the needed label, feel free to remove it if any concerns emerge.

/label docs-approved

File: Documentation/sample-metrics.md
Comment (lihongyan1) [inline Documentation/sample-metrics.md:16]: The query has issue, please refer doc pr
https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072

File: Documentation/telemetry/telemeter_query
Comment (lihongyan1) [inline Documentation/telemetry/telemeter_query:1]: The query has issue, please refer doc pr
https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072

File: assets/cluster-monitoring-operator/prometheus-rule.yaml
Comment (juzhao) [inline assets/cluster-monitoring-operator/prometheus-rule.yaml:407]: it seems this rule should define under openshift-logging prometheusrules file

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (juzhao) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:655]: FYI, these metrics in this PR are only added for telemeter client, you should also submit one PR to add these metrics to telemeter server
Docs on the process on why and how to send these metrics:
https://docs.google.com/document/d/1a6n5iBGM2QaIQRg9Lw4-Npj6QY9--Hpx3XYut-BrUSY/edit#heading=h.i05nm4wmvn7x

**Comment by juzhao** [comment]
FYI, these metrics in this PR are only added for telemeter client, you should also submit one PR to add these metrics to telemeter server
Docs on the process on why and how to send these metrics:
https://docs.google.com/document/d/1a6n5iBGM2QaIQRg9Lw4-Npj6QY9--Hpx3XYut-BrUSY/edit#heading=h.i05nm4wmvn7x

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (juzhao) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:624]: `log_logging_info  gives`
should remove one space after the metrics
`log_logging_info gives`
same for others

**Comment by pmoogi-redhat** [comment]
@juzhao removed prometheus recording rules for clo metrics and taken out quantile information from query and sample doc as per issue raised https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072. Please see if all issues raised by you got resolved. thanks.

**Comment by jan--f** [comment]
Seems like the generate test still has complaints. @pmoogi-redhat can you please double check that everything that a local call to `make generate` is committed here?

**Comment by pmoogi-redhat** [comment]
> Seems like the generate test still has complaints. @pmoogi-redhat can you please double check that everything that a local call to `make generate` is committed here?

Yes local call to make generate is committed however one review comment was to take out quantile from the query as stated by @juzhao so i made changes to take out quantile=~"0.99|0.99|0.99|" from query and documentation.

>>The query has issue, please refer doc pr
>>https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072

**Comment by pmoogi-redhat** [comment]
@juzhao and @jan--f  make generate in CI is complaining due to my taking out quantile information from telemetry query and generated documentation files.  Is there some other place that we require to take out quantile information as per the below earlier reported query issue, 
>> please refer doc pr https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072

File: Documentation/sample-metrics.md
Comment (pmoogi-redhat) [reply Documentation/sample-metrics.md]: @jan--f  please see https://github.com/openshift/openshift-docs/pull/42163#issuecomment-1049562072 here

**Comment by jan--f** [comment]
/lgtm
/label docs-approved
/label px-approved
Adding labels because a telemetry addition has no user or doc impact.
Ping @juzhao for qe-approved.

**Comment by pmoogi-redhat** [comment]
@jan--f resolved merge conflict. kindly see if this PR can be merged now.

**Comment by pmoogi-redhat** [comment]
resolved once again merge conflicts

**Comment by pmoogi-redhat** [comment]
@slashpai please review this PR asap

**Comment by pmoogi-redhat** [comment]
@jan--f earlier my PR didnt' get merged after getting lgtm from you as it was showing one review pending.  Post this a few PRs got merged and my PR again was showing merge conflicts. Could you please give me now lgtm  as i have resolved them yesterday ? Seeking you help to merge this PR as all changes already reviewed.

**Comment by slashpai** [comment]
@pmoogi-redhat  Due to rebase issue it was not merged earlier. Merge will happen automatically if all required tests are passing and lgtm label is there
/retest-required

File: Documentation/data-collection.md
Comment (juzhao) [inline Documentation/data-collection.md:718]: `log_logging_info  gives`
should remove one space after `log_logging_info`
`log_logging_info gives`

File: Documentation/data-collection.md
Comment (juzhao) [inline Documentation/data-collection.md:748]: metric is `cluster:log_logged_bytes_total:sum`

File: manifests/0000_50_cluster-monitoring-operator_04-config.yaml
Comment (juzhao) [inline manifests/0000_50_cluster-monitoring-operator_04-config.yaml:710]:
```
log_logging_info  gives
```
should remove one space after `log_logging_info`
```
log_logging_info gives
```
same for others

File: Documentation/data-collection.md
Comment (juzhao) [reply Documentation/data-collection.md:748]: please also fix the above issues in manifests/0000_50_cluster-monitoring-operator_04-config.yaml

File: Documentation/data-collection.md
Comment (pmoogi-redhat) [reply Documentation/data-collection.md:724]: it means no of failures occurred for collector pod deployment

File: Documentation/data-collection.md
Comment (juzhao) [reply Documentation/data-collection.md:718]: @pmoogi-redhat sorry, we have our rules to review the PR, accept the PR means we don't have concern for the PR, and it is easy to fix the issue

**Comment by juzhao** [comment]
@pmoogi-redhat thanks for the fix, and just find one issue
```
    # cluster:log_collected_bytes_total:sum gives total bytes collected by the collector and aggregated at each cluster level
    - '{__name__="cluster:log_collected_bytes_total:sum"}'
```
from the description, `cluster:log_collected_bytes_total:sum` and `cluster:log_logged_bytes_total:sum` are the same metrics
```
    # cluster:log_logged_bytes_total:sum gives total bytes collected by the collector and aggregated at each cluster level
    - '{__name__="cluster:log_logged_bytes_total:sum"}'
```

**Comment by pmoogi-redhat** [comment]
@juzhao corrected the description - please see now.

**Pull request merged**
